File: skills/excalibur-quest/scripts/quest_helper.js
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @fileoverview Cross-platform helper CLI for excalibur-quest remediation orchestrations.
 * Zero-dependency, compatible with Windows, macOS, and Linux Node.js environments.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

class QuestDB {
  constructor(dbDir) {
    this.dbDir = dbDir;
    if (!fs.existsSync(this.dbDir)) {
      fs.mkdirSync(this.dbDir, { recursive: true });
    }
    this.jsonDb = path.join(this.dbDir, 'quest_db.json');
    this.useDolt = this.checkDolt();
  }

  checkDolt() {
    try {
      execSync('dolt version', { stdio: 'ignore' });
      return true;
    } catch (e) {
      return false;
    }
  }

  loadQuest(issueId) {
    if (this.useDolt) {
      try {
        const query = `SELECT data FROM quests WHERE issue_id = '${issueId}'`;
        const res = execSync(`dolt sql -q "${query}" -r json`, { encoding: 'utf8' }).trim();
        if (res) {
          const rows = JSON.parse(res);
          if (rows && rows.rows && rows.rows.length > 0) {
            return JSON.parse(rows.rows[0].data);
          }
        }
      } catch (e) {
        console.error('Dolt load failed, falling back to JSON:', e.message);
      }
    }

    if (fs.existsSync(this.jsonDb)) {
      try {
        const data = JSON.parse(fs.readFileSync(this.jsonDb, 'utf8'));
        return data[issueId];
      } catch (e) {
        console.error('JSON load failed:', e.message);
      }
    }
    return null;
  }

  saveQuest(issueId, questData) {
    if (this.useDolt) {
      try {
        execSync('dolt sql -q "CREATE TABLE IF NOT EXISTS quests (issue_id VARCHAR(50) PRIMARY KEY, data TEXT)"', { stdio: 'ignore' });
        const escapedData = JSON.stringify(questData).replace(/'/g, "''");
        execSync(`dolt sql -q "REPLACE INTO quests (issue_id, data) VALUES ('${issueId}', '${escapedData}')"`, { stdio: 'ignore' });
        return;
      } catch (e) {
        console.error('Dolt write failed, falling back to JSON:', e.message);
      }
    }

    let data = {};
    if (fs.existsSync(this.jsonDb)) {
      try {
        data = JSON.parse(fs.readFileSync(this.jsonDb, 'utf8'));
      } catch (e) {}
    }
    data[issueId] = questData;
    try {
      fs.writeFileSync(this.jsonDb, JSON.stringify(data, null, 2), 'utf8');
    } catch (e) {
      console.error('JSON write failed:', e.message);
    }
  }
}

function acquireConcurrencySlot(lockDir, maxConcurrent = 3) {
  if (!fs.existsSync(lockDir)) {
    fs.mkdirSync(lockDir, { recursive: true });
  }

  while (true) {
    for (let i = 0; i < maxConcurrent; i++) {
      const lockPath = path.join(lockDir, `lock_${i}`);
      try {
        fs.mkdirSync(lockPath);
        console.error(`Acquired Roundtable seat ${i + 1} of ${maxConcurrent}`);
        return () => {
          try {
            fs.rmdirSync(lockPath);
          } catch (e) {}
        };
      } catch (err) {
        if (err.code !== 'EEXIST') {
          throw err;
        }
      }
    }
    console.error('Roundtable seats full. Waiting for slot...');
    const start = Date.now();
    while (Date.now() - start < 1000) {}
  }
}

function printHelpAndExit() {
  console.log(`
Excalibur Quest remediation coordinator.

Usage: node quest_helper.js <command> [options]

Commands:
  init   Initialize remediation beads database
  sync   Sync local beads state to GitHub
  close  Close beads and finalize quest

Options:
  --issue-id      Target GitHub issue number (required)
  --reviews-path  Path to JSON review findings file
  --repo-owner    GitHub repository owner (sync only)
  --repo-name     GitHub repository name (sync only)
  --output        Output JSON status file path (required)
`);
  process.exit(1);
}

function parseArgs() {
  const args = {};
  const argv = process.argv.slice(2);
  const command = argv[0];
  if (!command || !['init', 'sync', 'close'].includes(command)) {
    printHelpAndExit();
  }
  args.command = command;

  for (let i = 1; i < argv.length; i++) {
    if (argv[i].startsWith('--')) {
      const key = argv[i].slice(2).replace(/-([a-z])/g, (g) => g[1].toUpperCase());
      const val = argv[i + 1];
      if (val && !val.startsWith('--')) {
        args[key] = val;
        i++;
      } else {
        args[key] = true;
      }
    }
  }

  if (!args.issueId || !args.output) {
    console.error('Error: --issue-id and --output are required.\n');
    printHelpAndExit();
  }
  return args;
}

function runInit(args, db) {
  let beads = [];
  if (args.reviewsPath && fs.existsSync(args.reviewsPath)) {
    try {
      const reviews = JSON.parse(fs.readFileSync(args.reviewsPath, 'utf8'));
      if (reviews.findings) {
        beads = reviews.findings.map((r, idx) => ({
          id: `bead-${idx + 1}`,
          title: r.title || `Remediation item ${idx + 1}`,
          severity: r.severity || 'MEDIUM',
          status: 'PENDING'
        }));
      }
    } catch (e) {
      console.error('Failed to load reviews path:', e.message);
    }
  }

  if (beads.length === 0) {
    beads = [
      { id: 'bead-1', title: 'Audit source for security vulnerabilities', severity: 'CRITICAL', status: 'PENDING' },
      { id: 'bead-2', title: 'Verify tokio async I/O migrations', severity: 'HIGH', status: 'PENDING' },
      { id: 'bead-3', title: 'Check static analyzer schemas', severity: 'MEDIUM', status: 'PENDING' }
    ];
  }

  const questData = {
    issue_id: args.issueId,
    status: 'ACTIVE',
    beads: beads,
    created_at: Date.now() / 1000
  };
  db.saveQuest(args.issueId, questData);
  return questData;
}

function runSync(args, db) {
  const quest = db.loadQuest(args.issueId);
  if (!quest) {
    console.error(`Error: Quest for issue ${args.issueId} not initialized.`);
    process.exit(1);
  }

  if (!args.repoOwner || !args.repoName) {
    console.error('Error: --repo-owner and --repo-name are required for sync.');
    process.exit(1);
  }

  console.error(`Syncing beads status to GitHub repo ${args.repoOwner}/${args.repoName}...`);

  let commentBody = `### Excalibur Quest Sync Update\n\n**Quest Status:** ${quest.status}\n\n**Beads Progress:**\n`;
  for (const b of quest.beads || []) {
    const statusBox = b.status === 'COMPLETED' ? '[x]' : '[ ]';
    commentBody += `- ${statusBox} **${b.id}**: ${b.title} (${b.severity})\n`;
  }

  try {
    execSync(`gh issue comment ${args.issueId} --repo ${args.repoOwner}/${args.repoName} --body "${commentBody.replace(/"/g, '\\"')}"`, { stdio: 'pipe' });
    console.error('Successfully synced comment to GitHub.');
    quest.github_synced = true;
  } catch (e) {
    console.error('Warning: gh CLI sync failed (continuing locally):', e.message);
    quest.github_synced = false;
  }

  db.saveQuest(args.issueId, quest);
  return quest;
}

function runClose(args, db) {
  const quest = db.loadQuest(args.issueId);
  if (!quest) {
    console.error(`Error: Quest for issue ${args.issueId} not found.`);
    process.exit(1);
  }

  quest.status = 'COMPLETED';
  if (quest.beads) {
    quest.beads.forEach(b => b.status = 'COMPLETED');
  }

  db.saveQuest(args.issueId, quest);
  console.error(`Successfully closed quest ${args.issueId}.`);
  return quest;
}

function main() {
  const args = parseArgs();

  let workspaceRoot;
  try {
    workspaceRoot = execSync('git rev-parse --show-toplevel', { encoding: 'utf8' }).trim();
  } catch (e) {
    workspaceRoot = process.cwd();
  }

  const lockDir = path.join(workspaceRoot, '.excalibur_locks');
  const cleanup = acquireConcurrencySlot(lockDir);

  // Register cleanup exit hooks
  process.on('exit', cleanup);
  process.on('SIGINT', () => { cleanup(); process.exit(); });
  process.on('SIGTERM', () => { cleanup(); process.exit(); });

  try {
    const db = new QuestDB(path.join(workspaceRoot, '.excalibur_quest'));
    let result;

    if (args.command === 'init') {
      result = runInit(args, db);
    } else if (args.command === 'sync') {
      result = runSync(args, db);
    } else if (args.command === 'close') {
      result = runClose(args, db);
    }

    try {
      fs.writeFileSync(args.output, JSON.stringify(result, null, 2), 'utf8');
      console.log(`Success! Data written to: ${args.output}`);
    } catch (e) {
      console.error(`Error writing output to ${args.output}:`, e.message);
      process.exit(1);
    }
  } finally {
    cleanup();
  }
}

main();

File: skills/excalibur-quest/scripts/quest_helper.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Helper CLI for excalibur-quest remediation orchestrations."""

import argparse
import fcntl
import json
import os
import subprocess
import sys
import time


class QuestDB:
    """Quest tracking database with Dolt primary and local JSON fallback."""

    def __init__(self, db_dir):
        self.db_dir = db_dir
        os.makedirs(self.db_dir, exist_ok=True)
        self.json_db = os.path.join(self.db_dir, "quest_db.json")
        self.use_dolt = self._check_dolt()

    def _check_dolt(self):
        try:
            res = subprocess.run(
                ["dolt", "version"],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True
            )
            return res.returncode == 0
        except OSError:
            return False

    def load_quest(self, issue_id):
        if self.use_dolt:
            try:
                res = subprocess.run(
                    [
                        "dolt", "sql", "-q",
                        f"SELECT data FROM quests WHERE issue_id = '{issue_id}'",
                        "-r", "json"
                    ],
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    text=True
                )
                if res.returncode == 0 and res.stdout.strip():
                    rows = json.loads(res.stdout)
                    if rows.get("rows"):
                        return json.loads(rows["rows"][0]["data"])
            except Exception as e:
                print(f"Dolt load failed, falling back to JSON: {e}", file=sys.stderr)
        
        if os.path.exists(self.json_db):
            try:
                with open(self.json_db, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    return data.get(str(issue_id))
            except Exception as e:
                print(f"JSON load failed: {e}", file=sys.stderr)
        return None

    def save_quest(self, issue_id, quest_data):
        if self.use_dolt:
            try:
                subprocess.run(
                    [
                        "dolt", "sql", "-q",
                        "CREATE TABLE IF NOT EXISTS quests "
                        "(issue_id VARCHAR(50) PRIMARY KEY, data TEXT)"
                    ],
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE
                )
                escaped_data = json.dumps(quest_data).replace("'", "''")
                res = subprocess.run(
                    [
                        "dolt", "sql", "-q",
                        f"REPLACE INTO quests (issue_id, data) "
                        f"VALUES ('{issue_id}', '{escaped_data}')"
                    ],
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE
                )
                if res.returncode == 0:
                    return
            except Exception as e:
                print(f"Dolt write failed, falling back to JSON: {e}", file=sys.stderr)

        data = {}
        if os.path.exists(self.json_db):
            try:
                with open(self.json_db, "r", encoding="utf-8") as f:
                    data = json.load(f)
            except Exception:
                pass
        data[str(issue_id)] = quest_data
        try:
            with open(self.json_db, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=2)
        except OSError as e:
            print(f"JSON write failed: {e}", file=sys.stderr)


def acquire_concurrency_slot(lock_dir, max_concurrent=3):
    """Acquires a concurrency slot using file locks or blocks until one is free."""
    os.makedirs(lock_dir, exist_ok=True)
    locks = []
    for i in range(max_concurrent):
        lock_file = os.path.join(lock_dir, f"lock_{i}.lock")
        open(lock_file, "a").close()
        locks.append(open(lock_file, "r+"))

    while True:
        for idx, fh in enumerate(locks):
            try:
                fcntl.flock(fh, fcntl.LOCK_EX | fcntl.LOCK_NB)
                print(f"Acquired Roundtable seat {idx + 1} of {max_concurrent}", file=sys.stderr)
                return fh
            except (BlockingIOError, OSError):
                continue
        print("Roundtable seats full. Waiting for slot...", file=sys.stderr)
        time.sleep(1.0)


def run_init(args, db):
    """Initializes remediation quest with beads from review input."""
    beads = []
    if args.reviews_path and os.path.exists(args.reviews_path):
        try:
            with open(args.reviews_path, "r", encoding="utf-8") as f:
                reviews = json.load(f)
                for idx, r in enumerate(reviews.get("findings", [])):
                    beads.append({
                        "id": f"bead-{idx+1}",
                        "title": r.get("title", f"Remediation item {idx+1}"),
                        "severity": r.get("severity", "MEDIUM"),
                        "status": "PENDING"
                    })
        except Exception as e:
            print(f"Failed to load reviews path: {e}", file=sys.stderr)

    if not beads:
        beads = [
            {"id": "bead-1", "title": "Audit source for security vulnerabilities", "severity": "CRITICAL", "status": "PENDING"},
            {"id": "bead-2", "title": "Verify tokio async I/O migrations", "severity": "HIGH", "status": "PENDING"},
            {"id": "bead-3", "title": "Check static analyzer schemas", "severity": "MEDIUM", "status": "PENDING"}
        ]

    quest_data = {
        "issue_id": args.issue_id,
        "status": "ACTIVE",
        "beads": beads,
        "created_at": time.time()
    }
    db.save_quest(args.issue_id, quest_data)
    return quest_data


def run_sync(args, db):
    """Syncs beads database with GitHub issues."""
    quest = db.load_quest(args.issue_id)
    if not quest:
        print(f"Error: Quest for issue {args.issue_id} not initialized.", file=sys.stderr)
        sys.exit(1)

    print(f"Syncing beads status to GitHub repo {args.repo_owner}/{args.repo_name}...", file=sys.stderr)
    
    # Try using gh CLI to add comments on the GitHub issue
    comment_body = f"### Excalibur Quest Sync Update\n\n**Quest Status:** {quest['status']}\n\n**Beads Progress:**\n"
    for b in quest.get("beads", []):
        status_box = "[x]" if b["status"] == "COMPLETED" else "[ ]"
        comment_body += f"- {status_box} **{b['id']}**: {b['title']} ({b['severity']})\n"

    try:
        res = subprocess.run(
            [
                "gh", "issue", "comment", str(args.issue_id),
                "--repo", f"{args.repo_owner}/{args.repo_name}",
                "--body", comment_body
            ],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
        if res.returncode == 0:
            print("Successfully synced comment to GitHub.", file=sys.stderr)
            quest["github_synced"] = True
        else:
            print(f"Warning: gh CLI sync failed (continuing locally): {res.stderr.strip()}", file=sys.stderr)
            quest["github_synced"] = False
    except OSError:
        print("Warning: gh CLI not installed. Continuing local-only sync.", file=sys.stderr)
        quest["github_synced"] = False

    db.save_quest(args.issue_id, quest)
    return quest


def run_close(args, db):
    """Closes all beads and marks quest complete."""
    quest = db.load_quest(args.issue_id)
    if not quest:
        print(f"Error: Quest for issue {args.issue_id} not found.", file=sys.stderr)
        sys.exit(1)

    quest["status"] = "COMPLETED"
    for b in quest.get("beads", []):
        b["status"] = "COMPLETED"

    db.save_quest(args.issue_id, quest)
    print(f"Successfully closed quest {args.issue_id}.", file=sys.stderr)
    return quest


def main():
    parser = argparse.ArgumentParser(description="Excalibur Quest remediation coordinator.")
    subparsers = parser.add_subparsers(dest="command", required=True)

    # --- Subcommand: init ---
    p_init = subparsers.add_parser("init", help="Initialize remediation beads database")
    p_init.add_argument("--issue-id", required=True, type=int, help="Target GitHub issue number")
    p_init.add_argument("--reviews-path", help="Path to JSON review findings file")
    p_init.add_argument("--output", required=True, help="Output JSON status file path")

    # --- Subcommand: sync ---
    p_sync = subparsers.add_parser("sync", help="Sync local beads state to GitHub")
    p_sync.add_argument("--issue-id", required=True, type=int, help="Target GitHub issue number")
    p_sync.add_argument("--repo-owner", required=True, help="GitHub repository owner")
    p_sync.add_argument("--repo-name", required=True, help="GitHub repository name")
    p_sync.add_argument("--output", required=True, help="Output JSON status file path")

    # --- Subcommand: close ---
    p_close = subparsers.add_parser("close", help="Close beads and finalize quest")
    p_close.add_argument("--issue-id", required=True, type=int, help="Target GitHub issue number")
    p_close.add_argument("--output", required=True, help="Output JSON status file path")

    args = parser.parse_args()

    # Locate repository root / workspace directory to store locks and db safely
    try:
        workspace_root = subprocess.check_output(
            ["git", "rev-parse", "--show-toplevel"],
            stderr=subprocess.DEVNULL
        ).decode().strip()
    except subprocess.CalledProcessError:
        workspace_root = os.getcwd()

    lock_dir = os.path.join(workspace_root, ".excalibur_locks")
    lock_fh = acquire_concurrency_slot(lock_dir)

    try:
        db = QuestDB(os.path.join(workspace_root, ".excalibur_quest"))
        if args.command == "init":
            res = run_init(args, db)
        elif args.command == "sync":
            res = run_sync(args, db)
        elif args.command == "close":
            res = run_close(args, db)
        else:
            print(f"Unknown command: {args.command}", file=sys.stderr)
            sys.exit(1)

        # Write output file
        try:
            with open(args.output, "w", encoding="utf-8") as f:
                json.dump(res, f, indent=2)
            print(f"Success! Data written to: {args.output}")
        except OSError as e:
            print(f"Error writing output to {args.output}: {e}", file=sys.stderr)
            sys.exit(1)

    finally:
        lock_fh.close()


if __name__ == "__main__":
    main()
